dev_swift/05b_early_stopping.ipynb

```diff
@@ -80,65 +118,252 @@
+    "            if let recorder = callback as? Recorder { return recorder }\n",
+    "        }\n",
+    "        return nil\n",
+    "    }\n",
+    "}\n",
+    "\n",
+    "/// Hack\n",
+    "extension Learner {\n",
+    "    func makeTrainEvalDelegate() -> TrainEvalDelegate { return TrainEvalDelegate() }\n",
+    "}\n",
+    "\n",
+    "/// A non-generalized learning rate scheduler\n",
+    "extension Learner where O.Scalar == Float {\n",
+    "\n",
+    "    public class ParamScheduler: Delegate {\n",
+    "        public typealias ScheduleFunc = (Float) -> Float\n",
+    "\n",
+    "        // A learning rate schedule from step to float.\n",
+    "        public var scheduler: ScheduleFunc\n",
+    "        \n",
+    "        public init(scheduler: @escaping (Float) -> Float) {\n",
+    "            self.scheduler = scheduler\n",
+    "        }\n",
+    "        \n",
+    "        override public func batchWillStart(learner: Learner) {\n",
+    "            learner.optimizer.learningRate = scheduler(learner.pctEpochs/Float(learner.epochCount))\n",
     "        }\n",
-    "        print(loss)\n",
-    "        optimizer.update(&model[keyPath: variablesKeyPath], along: 𝛁model)\n",
     "    }\n",
     "}"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### LR Finder"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "NB: You may want to also add something that saves the model before running this, and loads it back after running - otherwise you'll lose your weights!"
+   ]
+  },
   {
    "cell_type": "code",
    "execution_count": null,
    "metadata": {},
    "outputs": [],
    "source": [
-    "// Example usage.\n",
-    "var model = Dense<Float>(inputSize: 784, outputSize: 10)\n",
-    "var optimizer = SGD<Dense<Float>, Float>(learningRate: 0.1)\n",
+    "func expSchedule(start: Float, end: Float, pct: Float) -> Float {\n",
+    "    return start * pow(end / start, pct)\n",
+    "}\n",
+    "\n",
+    "func makeAnnealer(start: Float, end: Float, schedule: @escaping (Float, Float, Float) -> Float) -> (Float) -> Float { \n",
+    "    return { pct in return schedule(start, end, pct) }\n",
+    "}\n",
     "\n",
-    "let data = Tensor<Float>(randomNormal: [10, 10, 784])\n",
-    "let labels = Tensor<Float>(randomNormal: [10, 10])\n",
-    "let dataset = Dataset<Example<Float, Float>>(elements: Example<Float, Float>(data: data, labels: labels))\n",
+    "/// LR Finder test\n",
+    "extension Learner where O.Scalar == Float {\n",
     "\n",
-    "train(&model, at: \\Dense<Float>.allDifferentiableVariables, on: dataset, using: &optimizer, loss: softmaxCrossEntropy)"
+    "    public class LRFinder2: Delegate {\n",
+    "        public typealias ScheduleFunc = (Float) -> Float\n",
+    "\n",
+    "        // A learning rate schedule from step to float.\n",
+    "        public var scheduler: ScheduleFunc\n",
+    "        public var numIter: Int\n",
```

Comment:
Swift naming conventions prefer the use of "count" instead of the prefix "num".

`numXX` -> `xxCount`
`numIter` -> `iterationCount`